### PR TITLE
Watch for the creation of missing files

### PR DESCRIPTION
**Pull request by amcasey**
When we discover that a file is missing, call sys.watchFile on its path so that we can update our state when it is created.

1) For tsc --watch, we schedule a new build
2) For tsserver, we update the containing projects

Caveat: VS will not know about the state change in tsserver until the next time it synchronizes its project list

### CODE REVIEW

**Comment by amcasey** [comment]
TODO: I'm still adding tests and cleaning things up, but I'd appreciate feedback on the general approach

**Comment by amcasey** [comment]
As far as I know, I've addressed all of @mhegazy's offline feedback.  I'll focus on testing until I hear otherwise.

**Comment by amcasey** [comment]
Added structure reuse tests and fixed bug where reused program structure dropped missing file list.

File: src/compiler/program.ts [hunk 1]
```diff
@@ -857,11 +860,31 @@ namespace ts {
...
+                    }
+                }
+            }
+
             // update fileName -> file mapping
             for (let i = 0; i < newSourceFiles.length; i++) {
                 filesByName.set(filePaths[i], newSourceFiles[i]);
             }
 
+            for (const p of oldProgram.getMissingFilePaths()) {
```
Comment (sheetalkamat) [inline src/compiler/program.ts]: this needs to be under the check of
```typescript
if (oldProgram.getMissingFilePaths) {
```

Comment (amcasey) [inline src/compiler/program.ts]: Oops :)

File: src/server/server.ts [hunk 2]
```diff
@@ -522,6 +522,9 @@ namespace ts.server {
                 return;
             }
 
+            // removed = deleted ? true : (added ? false : undefined)
```
Comment (amcasey) [inline src/server/server.ts]: It was intended as an explanation.  If you don't believe it's helpful, I'm happy to remove it.

File: src/compiler/program.ts (hunk 1 above)
Comment (amcasey) [inline src/compiler/program.ts]: I made the property optional in case there were existing implementers of the interface.  Was that over-cautious?

**Comment by amcasey** [comment]
I've posted a cleaned up PR: https://github.com/Microsoft/TypeScript/pull/16684.
@sheetalkamat: I made the change you suggested.
@rbuckton: I've responded to your questions above but not yet made any code changes.

**Pull request closed without merging**